RFC: drivers: modem: gsm_ppp: Optimizations and implement on/off

ycsin opened the pull request:
`gsm_ppp_start` and `gsm_ppp_stop` were introduced in #28475 that allows the modem to be turned on and off in the software level. After the merge of #34548, it is now possible to control the on and off of the modem in hardware level for the `gsm_ppp` modem.

This PR aims to:
1. cleanup some unused code,
2. changed how the driver is being compiled based on the Kconfig,
3. use a FSM for the setup, instead of the previous enum + boolean combo
4. further implement the `gsm_ppp_start` and `gsm_ppp_stop` so that it now turn off the hardware when the modem isn't in use to achieve power saving for battery operated devices.

fixes #33998

ycsin committed:
drivers: modem: gsm_ppp: Remove unused flags

The `setup_done` and `mux_setup_done` aren't being used
anywhere in the driver, therefore should be removed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Convert to use preprocessor IS_ENABLED

Convert `#if defined` and `if(IS_ENABLED` to `#if IS_ENABLED` so
that unused code will not be compiled and this also prevents
`#if defined` from being erroneously compiled if the CONFIG_FOO
is set to false.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Compilation optimization

Prevent code from being compiled if they are not used.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Reorder gsm mux logs so that it will be printed

The gsm_configure will return if it fails to perform mux_enable,
therefore the disable part of the log will not be printed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Remove unnecessary mux_enabled check

These mux_enabled checks aren't necessary because it will not
reach there if mux_enabled isn't true.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: modem_cmd_handler: Fix missing documentations

Fix missing documentations.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Implement on and off

Implemented gsm_ppp_start and gsm_ppp_stop to turn the modem on
or off to achieve power saving.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
May I have your comments please @mniestroj @jukkar @rlubos @nandojve

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -944,42 +965,48 @@ static void gsm_configure(struct k_work *work)
+		return;
+	}
+
+#if DT_INST_NODE_HAS_PROP(0, power_key)
```
#if DT_INST_NODE_HAS_PROP(0, power_src_gpios)

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1086,9 +1128,9 @@ static int gsm_init(const struct device *dev)
+#if CONFIG_GSM_PPP_AUTOSTART
+	gsm_ppp_start(dev);
+#endif
 
```
gsm->state = GSM_PPP_STOP;

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -930,6 +932,25 @@ static void gsm_configure(struct k_work *work)
+
+	if (gsm->state == GSM_PPP_PWR_SRC_ON) {
+		power_on_ops(&gsm->context);
+		gsm->state = GSM_PPP_WAIT_AT;
```
```c
		(void)k_work_reschedule(&gsm->gsm_configure_work, K_SECONDS(15));
		return;
```

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -893,7 +896,7 @@ static void mux_setup(struct k_work *work)
 		break;
 
-	case STATE_DONE:
+	case GSM_PPP_STATE_DONE:
```
Need to handle default case

ycsin commented:
@endian-benjamin Do you have time to review this?

endian-benjamin commented:
I'll have a look.

endian-benjamin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -829,13 +833,12 @@ static void mux_setup(struct k_work *work)
 	 */
-	if (IS_ENABLED(CONFIG_GSM_MUX) &&
-	    gsm->ppp_dev && gsm->state == STATE_CONTROL_CHANNEL) {
+	if (gsm->ppp_dev && gsm->state == GSM_PPP_STATE_CONTROL_CHANNEL) {
```
I don't think removing the ifdef wrap makes much sense, unless you explicitly verified that gcc is smart enough to realize `state == CONTROL_CHANNEL` is never true.

ycsin replied:
I don't really understand, are you referring to the removal of `IS_ENABLED(CONFIG_GSM_MUX)` in L832?
I removed it because `mux_setup` is only called if `CONFIG_GSM_MUX` is true so there isn't a need to check that again.

endian-benjamin replied:
Yeah. I missed that. Great!

endian-benjamin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -944,42 +968,48 @@ static void gsm_configure(struct k_work *work)
 	struct gsm_modem *gsm = dev->data;
 
+	if (gsm->state != GSM_PPP_STOP) {
+		LOG_ERR("gsm_ppp is already started");
```
Well now that you have a machinery for tracking state completely, changing the prototype of start/stop to return `int` would make sense imo. There's currently a problem where stopping gsm_ppp is hazarduous if the configuration loop is currently running - making start/stop return error codes would be a huge mitigation.

endian-benjamin commented on drivers/modem/gsm_ppp.h:
```diff
@@ -0,0 +1,108 @@
+ *
+ * @retval None.
+ */
+void enable_power_source(struct modem_context *ctx)
```
first of all these aren't namespaced appropriately.

endian-benjamin replied:
also they lack the inline qualifier?

ycsin replied:
Not very familiar with "inline", let me have a look online.

endian-benjamin commented:
Summary: 

- I really like the FSM for setup but it could be utilized better if you changed gsm_ppp_start/stop return error on illegal state transitions
- I think capturing all different modem power on/off sequences in this way would be tricky so personally I think the value of integrating power ops here is not that great. If gsm_ppp is supposed to be in charge of modem power, I would prefer generic implementations of `modem_power_on()` and off, but `__weak` linked so that the default behavior can be easily overridden by the app, without forcing the app developer to patch Zephyr.

ycsin commented:
@endian-benjamin Thanks for the review and comments!

ycsin committed:
drivers: modem: gsm_ppp: Unify retries variable

The attachment and the rssi readout are done in sequence so it
is possible to just recycle the same retries counter.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm_ppp: Customize error logs to ease debugging

Customize the error logs in each connection finalization steps
so that it is easier to trace the error without tapping the
modem's TX line.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

endian-benjamin commented:
> @endian-benjamin Thanks for the review and comments!
> 
> 1. Personally I don't have a usecase for the return value (yet), but I agree that the function can be more helpful to return error code instead of erring silently.
> 2. Do you mean separating the hw on/off out from the current `gsm_ppp_start/stop` into something like `gsm_ppp_power_on/off`? This sounds reasonably as well.
> 
> I think the having `__weak` sounds to be a good idea, but I'm not sure how it can be overridden outside of the `gsm_ppp`, especially when the `modem_context.h` is a private header and the `gsm_modem` is a private struct
> 
> Taking Quectel modem as an example, the reference manual states that to power off properly, user has to send AT command or press the power key and wait for the modem to reply "OK" before cutting off the power, that would at least require the `gsm_modem` struct to be public.

It would only require the modem cmd handler header to be public as that contains the modem_context struct. Other drivers already utilize that header (the modem shell). Though perhaps some cleanup would be needed to make it "presentable", so to speak.
 
> I do plan to expose that struct (some time later) so that it is possible to implement vendor-specific functions for the `gsm_ppp` in a separate header file which allows the `gsm_ppp` to be kept as generic as possible.

Well as you say yourself, Quectel requires weird solutions and many others do as well. And many custom PCBs will require weird solutions even if they run a "standard" modem, because some hw designer decided to skip some pins or whatever. 

Furthermore, power-on sequences called in the gsm_ppp setup loop has some implicit requirements - sleeping will cause the sys workq to get blocked, etc. All in all, the downsides to having gsm_ppp handle power-on becomes pretty apparent.

Personally, I'm not sure I see the huge upsides. What we do is usually something like `CONFIG_GSM_PPP_AUTOSTART=n`, then do the power-on sequence independently in the app or some other board-specific driver and THEN call `gsm_ppp_start()`. To restart the network stack, we just do

```
gsm_ppp_stop();
board_specific_modem_power_off();
board_specific_modem_power_on();
gsm_ppp_start();
```

Sure, this adds some boilerplate compared to letting gsm_ppp handle it under the hood, but I'm not convinced removing this boilerplate would be worth the downsides.

I'm open to discussion on it though - let's ping people. Perhaps this is even crucial enough to discuss at a developer meeting?

ycsin commented:
@endian-benjamin
> I'm open to discussion on it though - let's ping people. Perhaps this is even crucial enough to discuss at a developer meeting?

Not sure, since there's currently no maintainer for `gsm_ppp`

>Well as you say yourself, Quectel requires weird solutions and many others do as well. And many custom PCBs will require weird solutions even if they run a "standard" modem, because some hw designer decided to skip some pins or whatever.
>
>Furthermore, power-on sequences called in the gsm_ppp setup loop has some implicit requirements - sleeping will cause the sys workq to get blocked, etc. All in all, the downsides to having gsm_ppp handle power-on becomes pretty apparent.

Moving to `gsm_ppp` from the `bg9x` modem driver that I initially used, I always have the impression that the `gsm_ppp` is missing the dts key bindings that the `bg9x` has, so all this while from my perspective the `gsm_ppp` driver is missing the on/off functions. Only after reading your comments I realize that the power on/off operations can actually be done outside of the modem. I need to think about this again and see how I can modify the setup FSM and have a `_weak` empty on/off functions for people like me to overwrite in the application.

jukkar commented:
> Only after reading your comments I realize that the power on/off operations can actually be done outside of the modem.

It might be easier/simpler to do this modem specific reset/on/off thingy outside of the generic gsm_ppp code, unless you can come up with some solution that can be used by multiple modems.

ycsin commented:
> > Only after reading your comments I realize that the power on/off operations can actually be done outside of the modem.
> 
> It might be easier/simpler to do this modem specific reset/on/off thingy outside of the generic gsm_ppp code, unless you can come up with some solution that can be used by multiple modems.

What I know for Quectel modems is that they are more or less the same, just the period to hold the power key high/low is different. I will need to go through other existing modem drivers in Zephyr later.

endian-benjamin commented:
Note tho that even if I am not convinced regarding gsm_ppp handling power ctrl, I still think the setup FSM is a clear improvement. I would welcome that patch set for sure.

rlubos commented on drivers/modem/gsm_ppp.c:
```diff
@@ -587,7 +586,8 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		goto attaching;
 	}
 
-	if (IS_ENABLED(CONFIG_GSM_MUX) && gsm->mux_enabled) {
```
Why this change? We typically tend to include `IS_ENABLED` checks within C code instead of preprocessor when possible - it gives better code coverage during compilation. Same would go for some other places below.

rlubos commented on drivers/modem/gsm_ppp.c:
```diff
@@ -741,42 +746,42 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
+	int ret = 0;
 
 	/* Turn on muxing */
-	if (IS_ENABLED(CONFIG_MODEM_GSM_SIMCOM)) {
```
I'm not in favor of changes within this commit - as mentioned earlier, it's easier to introduce some regression if preprocessor checks are used and some particular configuration is not covered by the CI. IMO it's against what we typically do in Zephyr.

ycsin replied:
@rlubos meaning if I do `#if IS_ENABLED(CONFIG_GSM_MUX)` the mux related function will not be checked in CI?

rlubos replied:
It really depends on what samples are actually enabling gsm_ppp module and this particular feature (you'd need to analyze paritcualr samples and their sample.yaml files to see what's actually being built).

ycsin replied:
> It really depends on what samples are actually enabling gsm_ppp module and this particular feature (you'd need to analyze paritcualr samples and their sample.yaml files to see what's actually being built).

That means if we modify the `tests/drivers/build_all/modem`'s testcase.yaml it is possible to test these different features?

rlubos replied:
Build, not test. From what I see, that particular test project seems only to build some certain configurations, it does not have any test cases defined.

ycsin replied:
Is building them good enough?

rlubos replied:
It's the bare minimum I'd say.

ycsin replied:
But then we are compiling hundreds of lines of basically dead codes for users who aren't using the uart_mux and unused modem-specific's mux enabling steps in `mux_enable`

rlubos replied:
Yes, you're right, we're wasting some micro/milliseconds for the extra compilation. But we benefit from better compiler code coverage, and this approach is preferred in Zephyr:
https://docs.zephyrproject.org/latest/reference/util/index.html#c.IS_ENABLED

ycsin replied:
@rlubos got it, thanks!

ycsin commented:
I still think that it makes sense to do power control inside the driver so that it is more self-contained in the sense that the driver itself has the ability to fully control the hardware such as performing a power cycle if necessary (i.e. hardware not responding, max retries reached)

Having the turn on sequence in the `gsm_configure` makes it possible to retry the sequence if the hardware took too long to be AT ready.

I read a bit of how current modem being turned on in Zephyr:
1. hl7800 requires at least "pwr_on" and "reset" pin to asserted or deasserted
2. ublox sara has a "power pin" that needs to be held for a duration, reset seems to be optional
3. bg9x has a "power pin" that needs to be held for a duration
    - Same goes for other Quectel modems like EC2X/EG9X
    - My experience based on Quectel's evaluation board is that the reset pin itself is enough to turn on a modem, though power key/AT cmd is required to turn it off properly. From my experience, turning off the power works just as fine.
4. wnc_m14a2a requires quite a few more pins than power and reset

So seems like having a power key and reset pins are enough to turn on/off most of the modems supported in Zephyr. Power src pin for modem with power key is kind of a bonus to make sure that the modem is completely cutoff, it might be useful for modems that automatically boots up when powered and doesn't require a power key, however I'm not aware of any.

ycsin commented:
@rerickson1 May I have your comments?

bbilas commented:
> 
> My opinion would be to:
> 
>     * Expose `modem_context.h` so that it can be included in application
> 
>     * Implement `__weak` linked `gsm_mdm_on_ops` function that does whatever necessary for the modem to be AT ready.
>       
>       * Only power source, power key and maybe reset implementation are supported in `gsm_ppp`,
> 
>     * Implement `__weak` linked `gsm_mdm_off_ops` function that does whatever necessary to turn off the modem completely.
>       
>       * Power source and power key are supported in `gsm_ppp`
>       * If AT cmd is required to turn off a modem then we will have to expose `modem_cmd_handler.h` and have the `sem_response` as an argument so that it is possible to override this in application.

How about adding function callback within gsm_ppp_stop and gsm_ppp_start functions that would be called when those functions are executed? In that case the implementation of modem_power_on and modem_power_off is board-specific so it could be done outside gsm_ppp module. Maybe a global function to register global modem on/off callbacks would be even better.

jukkar commented:
> How about adding function callback within gsm_ppp_stop and gsm_ppp_start functions that would be called when those functions are executed? In that case the implementation of modem_power_on and modem_power_off is board-specific so it could be done outside gsm_ppp module. Maybe a global function to register global modem on/off callbacks would be even better.

That is a good proposal, just send patches if you have time to investigate this more.

wcappelle commented:
I'm testing the ppp with Quectell BG95, and for power on sequence I need to wait for a certain modem reply (AP READY) to be able to work properly in the initialization later on (just waiting for RDY makes one of the commands fail in init sequence). Therefore I think that everyone would benefit that the modem_cmd_handler functions would be public in that case?

ycsin commented:
> I'm testing the ppp with Quectell BG95, and for power on sequence I need to wait for a certain modem reply (AP READY) to be able to work properly in the initialization later on ...

It is an unsol output just like the "RDY" or you need to do something (send some AT command / trigger pin) for it to reply "AP READY"?

wcappelle commented:
unsol output indeed. 2 are comming, RDY & AP READY, but I need to wait for the second one to be able to continue with ppp start, otherwise the modem is not ready to receive on of the command (I think it was closing the current active connection(s)).

bbilas commented:
> > How about adding function callback within gsm_ppp_stop and gsm_ppp_start functions that would be called when those functions are executed? In that case the implementation of modem_power_on and modem_power_off is board-specific so it could be done outside gsm_ppp module. Maybe a global function to register global modem on/off callbacks would be even better.
> 
> That is a good proposal, just send patches if you have time to investigate this more.

Sure, but please clarify which approach you want me to send.

ycsin commented:
> unsol output indeed. 2 are comming, RDY & AP READY, but I need to wait for the second one to be able to continue with ppp start, otherwise the modem is not ready to receive one of the commands (I think it was closing the current active connection(s)).

So theoretically if I have a cmd handler that listen to "AP READY" and set a flag high, then in my `gsm_configure` work I just need to make sure that flag is high before I proceed to do the "AT" and everything else?

> Sure, but please clarify which approach you want me to send.

By "just send patches" he means submitting PR containing what you have suggested.

bbilas commented:
> > Sure, but please clarify which approach you want me to send.
> 
> By "just send patches" he means submitting PR containing what you have suggested.

Heh, I know but I've provided two different proposals so I don't know which one should I send.

jukkar commented:
> Heh, I know but I've provided two different proposals so I don't know which one I should send.

Hmm, I thought there was only one proposal :)

Anyway, setting up callbacks via some register functions sounds like a good proposal to me.

ycsin commented:
Closing this, since #38946 is prefered and remaining changes already in #38596
